akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala

```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
 
     private var currentRoutees: Array[ActorRef[T]] = _
 
-    override def selectRoutee(): ActorRef[T] = {
+    override def selectRoutee(msg: T): ActorRef[T] = {
       val selectedIdx = ThreadLocalRandom.current().nextInt(currentRoutees.length)
       currentRoutees(selectedIdx)
     }
+
     override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
       currentRoutees = newRoutees.toArray
     }
+  }
 
+  final class ConsistentHashingLogic[T](
+      virtualNodesFactor: Int,
+      mapping: ConsistentHashMapping[T],
+      system: ActorSystem[T])
+      extends RoutingLogic[T] {
+    require(system ne null, "system argument of ConsistentHashingLogic cannot be null.")
+
+    private var currentRoutees: Set[ActorRef[T]] = Set.empty
+
+    private var consistentHash: ConsistentHash[ActorRef[T]] = ConsistentHash(currentRoutees, virtualNodesFactor)
+
+    override def selectRoutee(msg: T): ActorRef[T] =
+      try {
+        mapping(msg) match {
+          case bytes: Array[Byte] => consistentHash.nodeFor(bytes)
+          case str: String        => consistentHash.nodeFor(str)
+          case x: AnyRef =>
+            val bytes = SerializationExtension(system.toClassic).serialize(x).get
+            consistentHash.nodeFor(bytes)
+        }
+      } catch {
+        case NonFatal(e) =>
+          system.log.warn("Couldn't route message [{}] due to [{}]", msg, e.getMessage)
+          system.deadLetters
+      }
```

Comment:
Let thrown errors pass through to the actor. The router actor will never even ask if the list of routees is empty, so this can only be a hash function error. Not 100% sure if we should handle it like this or rather let the router fail instead (and then you can supervise the router).